README.md
# Iugu WHMCS Pro

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/4ab6dcf71d8946b892cfd8749f9666ee)](https://www.codacy.com/app/andrekutianski/iugu-whmcs-pro?utm_source=github.com&utm_medium=referral&utm_content=eunarede/iugu-whmcs-pro&utm_campaign=badger)
[![GitHub issues](https://img.shields.io/github/issues/eunarede/iugu-whmcs-pro.svg?style=flat-square)](https://github.com/eunarede/iugu-whmcs-pro/issues)
[![GitHub forks](https://img.shields.io/github/forks/eunarede/iugu-whmcs-pro.svg?style=flat-square)](https://github.com/eunarede/iugu-whmcs-pro/network)
[![GitHub stars](https://img.shields.io/github/stars/eunarede/iugu-whmcs-pro.svg?style=flat-square)](https://github.com/eunarede/iugu-whmcs-pro/stargazers)
[![Twitter](https://img.shields.io/twitter/url/https/github.com/eunarede/iugu-whmcs-pro.svg?style=social&style=flat-square)](https://twitter.com/intent/tweet?text=Wow:&url=%5Bobject%20Object%5D)
[![Gitter](https://badges.gitter.im/eunarede/iugu-whmcs-pro.svg)](https://gitter.im/eunarede/iugu-whmcs-pro?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)


Este módulo foi últil para você? Mostre sua gratidão contribuindo com um choop, café ou até mesmo um salto de paraquedas!


* [Requerimentos](https://github.com/eunarede/iugu-whmcs-pro/wiki/Requerimentos)
* [Instalação](https://github.com/eunarede/iugu-whmcs-pro/wiki/instalacao)

O Módulo para [WHMCS] Iugu Pro desenvolvido pela [EunaRede] proporciona uma integração completa e transparente com o gateway de pagamento [Iugu]

Este módulo consiste em dois métodos de pagamento diferentes, boleto bancário e cartão de crédito, confira os recursos a seguir:

## Módulo Boleto Bancário

![Boleto Bancário](docs/img/frontend-bank_slip.png)

Através do método de pagamento por boleto bancário do Módulo [WHMCS] Iugu Pro, é possível gerar o boleto diretamente no WHMCS, sem necessidade de redirecionamento.

### Com este módulo você poderá:

- [x] Gerar o boleto sem redirecionar o cliente para o site da Iugu;
- [x] Dar baixa na fatura automaticamente após a compensação (retorno automático);
- [x] Cancelar fatura na Iugu assim que cancelada no WHMCS;
- [x] Gerar segunda via de boleto caso fatura esteja vencida automaticamente;
- [x] Inserir multa na segunda via do boleto automaticamente;
- [ ] ~~Cadastrar o cliente do [WHMCS] na Iugu automaticamente~~;
- [ ] ~~Excluir o cliente na Iugu quando excluido no [WHMCS]~~;

## Módulo Cartão de Crédito (experimental)

![Tela do cartão de crédito](docs/img/frontend-credit-card.png)

Assita um video demonstrando as funcionalidades do módulo cartão de crédito em [https://youtu.be/00dDINuXT1g](https://youtu.be/00dDINuXT1g)

Através do método de pagamento por cartão de crédito do Módulo [WHMCS] Iugu Pro, é possível realizar o recebimento de faturas via cartão de crédito diretamente no [WHMCS], sem necessidade de redirecionamento. O módulo utiliza os campos originais de cartão de crédito do [WHMCS] sem necessidade de modificação no tema.

Ao realizar um pedido ou na atualização do cartão de crédito no perfil de um cliente já existente, o módulo captura os dados do cartão e utilizando a API da Iugu, gera um código único criptogrado: o token de pagamento.

Os dados do cartão de crédito são criptografados e armazenados nos servidores da Iugu, gerando um token de representação destes dados, que são atrelados ao cliente dentro do sistema da Iugu. Este token então é armazenado no [WHMCS] e vinculado a conta do cliente no [WHMCS].

Através do token de pagamento, o [WHMCS] poderá realizar capturas automáticas de pedidos e cobranças recorrentes de faturas no cartão de crédito. O [WHMCS] armazena apenas a data de vencimento do cartão, os 4 últimos digitos e bandeira (estas informações são utilizadas para alertas de vencimento do cartão).

### Com este módulo você poderá:

- [ ] ~~Realizar a cobrança da fatura sem redirecionar o cliente para o site da Iugu~~;
- [ ] ~~Capturar os cartão de crédito de forma transparente~~;
- [ ] ~~Capturar faturas recorrentes automaticamente~~;
- [x] Cadastrar o cliente do [WHMCS] na Iugu automaticamente;
- [ ] ~~Excluir o cliente na Iugu quando excluido no [WHMCS]~~;
- [x] Atualizar os dados do cartão do cliente diretamente no [WHMCS];

## Atualização do Cartão de Crédito

Atualização do cartão de crédito do cliente diretamente no [WHMCS]. Ao atualizar o cartão de crédito de um cliente, será gerado novo Token de cobrança recorrente na Iugu. Nunca mais perca uma venda e evite que seu cliente esqueça de pagar suas faturas na data de vencimento!

Assista o [vídeo de demonstração](https://youtu.be/dqzsNcbA-Gg).

![Atualização do cartão](https://raw.githubusercontent.com/wiki/eunarede/iugu-whmcs-pro/imgs/giphy.gif)

## Instalação

Estamos constantemente aprimorando este módulo e enriquecendo sua documentação. Para acessar a documentação de instalação, veja nossa [Wiki]

## Requerimentos

Veja os requerimentos necssários para este módulo acessando nossa [Wiki]

## Créditos e isenções

Este módulo é entregue "como está", a EunaRede ou sua equipe não se responsabilizam por qualquer eventual projuízo ou perda de qualquer natureza que este plugin ou a sua utilização possam estar associados.

A EunaRede e nenhum de seus membros estão associados ou vinculados de alguma maneira com a Iugu. Este desenvolvido é de livre iniciativa e independente.

[EunaRede]: https://www.eunarede.com
[Iugu]: https://iugu.com
[WHMCS]: https://www.whmcs.com/members/aff.php?aff=4571
[Wiki]: https://github.com/eunarede/iugu-whmcs-pro/wiki


changelog.md
# 1.6.4
+ Correções na busca de faturas
# 1.6.3
+ Correções no tratamento de erros
# 1.6.2
+ Tratamento de retorno da API Iugu para geração de de dados para o "System Module Debug Log" do WHMCS, facilitando a identificação de problemas.
# 1.6.1
+ tratativa de erros.
# 1.6.0
+ implementada a geração de segunda via da fatura caso a atual esteja vencida.
+ implementada mensagem de erro quando fatura expirada;
+ removida a geração de clientes da Iugu.
+ Removida a biblioteca Iugu PHP.
+ Implementada a comunicação com a API através da biblioteca [Unirest PHP](https://github.com/Mashape/unirest-php).
+ +implementado gatilho para cancelamento da fatura Iugu caso a fatura vinculada no WHMCS for cancelada.
# 1.5.9
+ [fix] Inserido o parametro _number_ para cadastramento do cliente na Iugu para constante '000'. API Iugu exige que ao enviar o parametro _zip_code_ seja enviado _number_, WHMCS não possui este parametro, o que ocasionava problemas no cadastramento do cliente como mencionado na issue #14
# 1.5.8
+ corrigido erro 500 quando a fatura está vencida. Caso esteja a data do boleto é alterada para o dia atual #14
+ adicionado o campo personalizado de CPF/CNPJ para cadastramento na Iugu para atender a nova exigencia para boletos registrados #9
+ corrigido o problema de cadastramendo do cliente na Iugu quando é alterado a forma de pagamento entre cartão/boleto boleto/cartão #12 #14
# 1.5.7
+ Remoção do submodulo do SDK da Iugu.
+ Correção do diretório do SDK Iugu.
+ Provavelmente resolva o problema reportado aqui #12
# 1.5.3.1
+ correção de erro 505 após ativar o módulo #5
+ aprimoramento dos dados enviados a Iugu para cadastro do cliente #6
# v1.5.3
+ remoção de campo de usuário API e atualização da documentação
# v1.5.2
+ tratamento de geração de faturas no Painel Iugu para evitar duplicidades de faturas para a mesma Invoice do WHMCS
+ Criação de tabela adicional e módulo addon para ativação
+ pequenas melhorias nos tratamentos de erros
# v1.5.1
+ Inserção de variável para pagamento via boleto bancário;
# v1.5
+ Separação dos meios de pagamentos;
+ Tratamento de todas as variaveis para normatização com WHMCS v6.x;
+ Tratamento do retorno dos pagamentos de formas individuais (CC e Boleto);


modules/addons/iugu/iugu.php
<?php

// http://docs.whmcs.com/Addon_Module_Developer_Docs

/**
 *
 * @ IUGU ADDON FOR WHMCS
 *
 * @ Version  : 7.X
 * @ Author   : EUNAREDE
 * @ Release on : 2016-10-21
 * @ Website  : http://www.eunarede.com
 *
 * */

if (!defined("WHMCS"))
    die("Esse arquivo não pode ser acessado diretamente.");

use Illuminate\Database\Capsule\Manager as Capsule;

function iugu_config() {
    $configarray = array(
		"name" => "Iugu",
		"description" => "Addon necessário para o funcionamento do módulo de pagamento da Iugu.",
		"version" => "2.0",
		"author" => "EunaRede",
    "fields" => array(
                      "iugu_account_number" => array(
                                                    "FriendlyName" => "Número da Conta Iugu",
                                                    "Type" => "text",
                                                    "Size" => "25",
                                                    "Description" => "O número da sua conta Iugu (não confundir com api)",
                                                    ),
                      "iugu_api_token" => array(
                                              "FriendlyName" => "Token da API Iugu",
                                              "Type" => "text",
                                              "Size" => "25",
                                              "Description" => "A Chave de API (Token)",
                      ),
    ),
	);
    return $configarray;
}


function iugu_activate() {

    #
    # Cria as tabelas customizadas para armazenar as faturas emitidas e as associações com os ID's dos clientes
    # cadastrados na Iugu
    #
    $query = "CREATE TABLE `mod_iugu_invoices` (`id` INT(10) NOT NULL AUTO_INCREMENT PRIMARY KEY, `invoice_id` int(11) NOT NULL, `iugu_id` varchar(255) NOT NULL, `secure_id` varchar(255) NOT NULL)";
    $result = mysql_query($query);
    $query = "CREATE TABLE `mod_iugu_customers` (`id` INT( 10 ) NOT NULL AUTO_INCREMENT PRIMARY KEY, `user_id` int(11) NOT NULL, `iugu_id` varchar(255) NOT NULL)";
    $result = mysql_query($query);

    # Return Result
    return array('status'=>'success','description'=>'Addon instalado com sucesso! Você já pode configurar/utilizar o módulo de pagamento da Iugu.');
    return array('status'=>'error','description'=>'Erro ao instalar addon.');

}

function iugu_output($vars){

  require_once("iugu-php/lib/Iugu.php");

  $iuguApiToken = $vars['iugu_api_token'];

  // Busca na tabela mod_iugu se já existe uma fatura criada na Iugu referente a invoice do WHMCS
  //$iuguInvoiceId = Array();
  try{
  // $iuguInvoiceId = Capsule::table('mod_iugu')->where('invoice_id', $invoiceId)->value('iugu_id');
  $tableUsers = Capsule::table('mod_iugu_customers')->select('user_id', 'iugu_id')->get();
}catch (\Exception $e){
  echo "Problemas em localizar a sua fatura. Contate nosso suporte e informe o erro 001. {$e->getMessage()}";
}
echo "<div class='tablebg'>";
echo '<h2>Usuários com Cadastro na Iugu</h2><p>Usuários do WHMCS com cadastro associado na Iugu. Dados do Banco</p>';
echo "<table id='sortabletbl0' class='datatable' width='100%' border='0' cellspacing='1' cellpadding='3'>";
echo '<thead>';
echo '<tr>';
echo '<th>User</th>';
echo '<th>Iugu User</th>';
echo '</tr>';
echo '</thead><tbody>';
echo '<tbody>';
foreach ($tableUsers as $key => $value) {
  echo '<tr><td>';
  print_r($value->user_id);
  echo '</td><td>';
  print_r($value->iugu_id);
  echo '</td></tr>';
}
echo '</tbody></table></div>';

try{

  Iugu::setApiKey($iuguApiToken);
  $customers = Iugu_Customer::search()->results();
  // var_dump($customers);
  echo "<div class='tablebg'>";
  echo '<h2>Usuários Iugu</h2><p>Usuários cadastrados na Iugu. Dados da API</p>';
  echo "<table id='sortabletbl0' class='datatable' width='100%' border='0' cellspacing='1' cellpadding='3'>";
  echo '<thead>';
  echo '<tr>';
  echo '<th>Nome</th>';
  echo '<th>Email</th>';
  echo '<th>ID</th>';
  echo '</tr>';
  echo '</thead><tbody>';
  echo '<tbody>';
  foreach ($customers as $key => $value) {
    echo '<tr><td>';
    print_r($value->name);
    echo '</td><td>';
    print_r($value->email);
    echo '</td><td>';
    print_r($value->id);
    echo '</td></tr>';
  }
  echo '</tbody></table></div>';

}catch (\Exception $e){

  echo "Problemas em localizar a sua fatura. Contate nosso suporte e informe o erro 001. {$e->getMessage()}";
}

}



function iugu_deactivate() {

    # Remove Custom DB Table
    $query = "DROP TABLE `mod_iugu_invoices`";
    $result = mysql_query($query);
    $query = "DROP TABLE `mod_iugu_customers`";
    $result = mysql_query($query);

    # Return Result
    return array('status'=>'success','description'=>'Addon desinstalado com sucesso. Não deixei de desativar o módulo de pagamento da Iugu para evitar problemas futuros.');
    return array('status'=>'error','description'=>'Erro ao desinstalar addon.');

}


modules/addons/iugu/hooks.php
<?php
/**
 *
 * @ IUGU HOOKS FOR WHMCS
 *
 * @ Version  : 7.X
 * @ Author   : EUNAREDE
 * @ Release on : 2016-10-21
 * @ Website  : http://www.eunarede.com
 *
 * */
if (!defined( "WHMCS" )) {
	exit( "This file cannot be accessed directly" );
}
use Illuminate\Database\Capsule\Manager as Capsule;
use Unirest\Request as Request;
use Unirest\Body as Body;
// require_once("src/Unirest.php");

# Remove dados do cliente no sistema Iugu quando o mesmo for removido do WHMCS

function hook_delete_iugu_client( $vars ) {

	require_once("iugu-php/lib/Iugu.php");

	// Dados para conexão com API da Iugu
	// $addonData = Capsule::table('tbladdonmodules')->where('module', 'iugu')->get();
	// $apiToken = $addonData['iugu_api_token'];
	// $accountNumber = $addonData['iugu_account_number'];

	$table = "tbladdonmodules";
	$fields = "value";
	$where = array("module" => "iugu","setting" => "iugu_api_token");
	$result = select_query($table,$fields,$where);
	$data = mysql_fetch_array($result);

	$apiToken = $data["value"];

	// Dados do cliente no WHMCS para exlcusão
	$userid = $vars['userid'];
	$iuguUserId = Capsule::table('mod_iugu_customers')->where('user_id', $userid)->value('iugu_id');
	// se os dados existirem na tabela mod_iugu_customers então busque na Iugu e apague
	if($iuguUserId) {
		Iugu::setApiKey($apiToken);
		$customer = Iugu_Customer::fetch($iuguUserId);
		$customer->delete();
		// remove da tabela local os dados do cliente associado a Iugu
		try{
			Capsule::table('mod_iugu_customers')->where('user_id', $userid)->delete();
		}catch(\Exception $e){
			echo "Problemas em apagar o cliente no banco de dados local. {$e->getMessage()}";
		}
		$response = json_decode($customer, true);
		logModuleCall("Iugu Clients","Apagar Cliente",$iuguUserId,$response);
}

}

function hook_cancel_invoice_iugu( $vars ) {

// use Unirest\Request as Request;
// use Unirest\Body as Body;
require_once("src/Unirest.php");


	$apitoken = Capsule::table('tbladdonmodules')->where([
		['module', '=', 'iugu'],
		['setting', '=', 'iugu_api_token'],
	])->value('value');

	$iuguinvoiceid = Capsule::table('mod_iugu_invoices')
									->where('invoice_id', $vars['invoiceid'])
									->value('iugu_id');

									if (!is_null($iuguinvoiceid)){
										Unirest\Request::auth("$apitoken", '');
										$result = Unirest\Request::put("https://api.iugu.com/v1/invoices/$iuguinvoiceid/cancel");
									}




}
//add_hook('PreDeleteClient', 1, 'hook_delete_iugu_client');
// add_hook('ClientAdd', 1, 'hook_create_client');
add_hook('InvoiceCancelled', 1, 'hook_cancel_invoice_iugu');
 ?>


modules/gateways/iugu_cartao.php
<?php
if (!defined("WHMCS")) {
    die("This file cannot be accessed directly");
}
use Illuminate\Database\Capsule\Manager as Capsule;
require_once("iugu/Iugu.php");
/**
 * Define module related meta data.
 *
 * Values returned here are used to determine module related capabilities and
 * settings.
 *
 * @see http://docs.whmcs.com/Gateway_Module_Meta_Data_Parameters
 *
 * @return array
 */
function iugu_cartao_MetaData(){
    return array(
        'DisplayName' => 'Iugu WHMCS Pro - Cartão',
        'APIVersion' => '1.1', // Use API Version 1.1
        'DisableLocalCredtCardInput' => true,
        'TokenisedStorage' => true,
    );
}

/**
 * Define gateway configuration options.
 *
 * The fields you define here determine the configuration options that are
 * presented to administrator users when activating and configuring your
 * payment gateway module for use.
 *
 * Supported field types include:
 * * text
 * * password
 * * yesno
 * * dropdown
 * * radio
 * * textarea
 *
 * Examples of each field type and their possible configuration parameters are
 * provided in the sample function below.
 *
 * @return array
 */
function iugu_cartao_config(){
    return array(
        // the friendly display name for a payment gateway should be
        // defined here for backwards compatibility
        'FriendlyName' => array(
            'Type' => 'System',
            'Value' => 'Iugu WHMCS Pro - Cartão',
        ),
        // a text field type allows for single line text input
        'iugu_account_id' => array(
            'FriendlyName' => 'Número da Conta',
            'Type' => 'text',
            'Size' => '40',
            'Default' => '',
            'Description' => 'Número da sua conta Iugu. Não confunda com API.',
        ),
        'api_token' => array(
            'FriendlyName' => 'Token',
            'Type' => 'text',
            'Size' => '40',
            'Default' => '',
            'Description' => 'Acesse sua conta Iugu para gerar seu token.',
        ),
        'cpf_cnpj_field' => array(
            'FriendlyName' => 'Campo CPF/CNPJ',
            'Type' => 'text',
            'Size' => '20',
            'Default' => '',
            'Description' => 'Insira o nome referente ao campo CPF/CNPJ',
        ),
        'test_mode' => array(
            'FriendlyName' => 'Ativar modo teste',
            'Type' => 'yesno',
            'Description' => 'ativar a utilização de cartões de crédito de teste.'
        ),
    );
}

// Cadastra o cliente na Iugu com os dados disponíveis no WHMCS
function iugu_cartao_add_client( $params ) {

  $campoDoc = $params['cpf_cnpj_field'];

  Iugu::setApiKey($params['api_token']);

  $iuguUser = Iugu_Customer::create(Array(
    "email" => $params['clientdetails']['email'],
    "name" => $params['clientdetails']['fullname'],
    "notes" => "Cliente cadastrado através do WHMCS",
    "cpf_cnpj" => $params['clientdetails'][$campoDoc],
    "zip_code" => $params['clientdetails']['postcode'],
    "number" => "000",
    "custom_variables" => Array(
      Array(
        "name" => "whmcs_user_id",
        "value" => $params['clientdetails']['userid']
      ))
  ));
  // Insere na tabela mod_iugu_customers o Código do cliente Iugu
  Capsule::table('mod_iugu_customers')->insert(
                                    [
                                      'user_id' => $params['clientdetails']['userid'],
                                      'iugu_id' => $iuguUser->id
                                    ]
                                  );
  logModuleCall("Iugu","iugu_create_client",$params['clientdetails']['userid'],$iuguUser);
  return $iuguUser->id;
}

// Busca na tabela mod_iugu_customers se já existe o cliente cadastrado
function iugu_cartao_search_client( $user ) {

    // procura no banco
    $iuguUserId = Capsule::table('mod_iugu_customers')->where('user_id', $user)->value('iugu_id');

    // loga a ação para debug
    logModuleCall("Iugu","iugu_search_client",$user,$iuguUserId);

    // retorna o ID do cliente
    return $iuguUserId;

}


function iugu_cartao_storeremote($params){

  $apiToken = $params['api_token'];
  $testMode = "true";
  if (!$params['test_mode']) {
    $testMode = "false";
  }
  $accountId = $params['iugu_account_id'];
  // parametro com o ID da conta WHMCS
  $userid = $params['clientdetails']['userid'];
  // Dados do cliente recebidos como sugerido pela equipe do WHMCS
  $clientFirstName = $params['clientdetails']['firstname']; # cliente first name
  $clientLastName = $params['clientdetails']['lastname']; # client last name
  $clientEmail = $params['clientdetails']['email']; # client email
  // tratamento do vencimento do cartão de crédito para a Iugu
  $cardMonth = substr( $params['cardexp'], 0, -2 ); # trata a expiração do cc para mês
  // $cardYear = substr_replace('20', $params['cardexp'], 2, 0); # trata a expiração do cc para ano
  $cardYear = substr( $params['cardexp'], 2, 2 ); # trata a expiração do cc para ano
  // Dados do cartão
  $cardNumber = $params['cardnum']; # the Card Number
  $cardCvv = $params['cardcvv']; # the verification card code
  // Busca na tabela mod_iugu se já existe um cliente criado na Iugu
  try{
  $iuguClientId = iugu_cartao_search_client( $userid );
  if ( is_null($iuguClientId) ) {
    $iuguClientId = iugu_cartao_add_client($params);
  }
  logModuleCall("iugu_cartao_storeremote","Buscar Cliente",$userid,$iuguClientId);
  }catch (\Exception $e){
  //logModuleCall("Iugu Cartao","Buscar Cliente",$userid,$iuguClientId);
  echo "Problemas em localizar o cliente no banco de dados local. Erro 001. {$e->getMessage()}";
  }

  // Cria o token de pagamento através dos dados do cartão do cliente
  $urlToken = 'https://api.iugu.com/v1/payment_token';
  $postfieldsToken = array(
    "account_id" => $accountId,
    "method" => 'credit_card',
    "test" => $testMode,
    "data[number]" => $cardNumber,
    "data[verification_value]" => $cardCvv,
    "data[first_name]" => $clientFirstName,
    "data[last_name]" => $clientLastName,
    "data[month]" => $cardMonth,
    "data[year]" => $cardYear,
  );

  $responseToken = curlCall($urlToken, $postfieldsToken);
  $iuguCreditCardToken = json_decode($responseToken, true);

  logModuleCall("Iugu Cartao","Criar Token",$postfieldsToken,$iuguCreditCardToken);

    //cria o método de pagamento padrão para armazenar o token do cartão de credito na iugu
    $urlMethod = "https://api.iugu.com/v1/customers/{$iuguClientId}/payment_methods";
    $postfieldsMethod = array(
      "api_token" => $apiToken,
      "description" => 'Cartão Inserido via WHMCS',
      "token" => $iuguCreditCardToken['id'],
      "set_as_default" => "true",
    );

    $payment_method = curlCall($urlMethod, $postfieldsMethod);
    $paymentMethodToken = json_decode($payment_method, true);

    logModuleCall("Iugu Cartao","Criar Metodo de Pgto",$payment_method,$paymentMethodToken);

    //se o metodo de pagamento foi criado com sucesso, retorna sucesso ao _storeremote
    if($paymentMethodToken['id']){
      return array(
        "status" => "success",
        "gatewayid" => $paymentMethodToken['id'],
        "rawdata" => $paymentMethodToken,
      );
    }
    //senão retorna falha ao _storeremote
    else{
      return array(
        "status" => "failed",
        "rawdata" => $paymentMethodToken,
      );
    }


} //iugu_cartao_storeremote


function iugu_cartao_capture( $params ){

  // System Parameters
	$apiToken = $params['api_token'];

  $userid = $params['clientdetails']['userid'];
  //var_dump($cpf_cnpj);

	// Invoice Parameters
	$invoiceId = $params['invoiceid'];
  $tokenisedPaymentId = $params['gatewayid'];

	/** @var stdClass $client */
	$itens = Array();
	try {
    $selectInvoiceItens = Capsule::table('tblinvoiceitems')->select('amount', 'description')->where('invoiceid', $invoiceId)->get();
    logModuleCall("Iugu Cartao","Buscar Itens da Fatura","Itens",$selectInvoiceItens);
			}catch (\Exception $e) {
    		echo "Não foi possível gerar os itens da fatura. Erro 002 {$e->getMessage()}";
				}

  foreach ($selectInvoiceItens as $key => $value) {
    $valor = number_format($value->amount, 2, '', '');
    $item = Array();
    $item['description'] = $value->description;
    $item['quantity'] = "1";
    $item['price_cents'] = $valor;
    $itens[] = $item;
  }

  // Busca na tabela mod_iugu_customers o ID do cliente da Iugu
  try{
    $iuguCustomerId = iugu_cartao_search_client( $userid );
    logModuleCall("Iugu","iugu_cartao_capture","Cliente",$iuguCustomerId);
  }catch (\Exception $e){
    echo "Problemas em localizar o cliente no banco de dados local. {$e->getMessage()}";
  }

  if (!$iuguCustomerId) {
    $iuguCustomerId = iugu_cartao_add_client( $params );
  }

	Iugu::setApiKey($apiToken);
	$chargeInvoice = Iugu_Charge::create(Array(
    "customer_payment_method_id" => $tokenisedPaymentId,
    "customer_id" => $iuguCustomerId,
     "email" => $email,
		"items" => $itens
	));

 if($chargeInvoice->success){
   logModuleCall("Iugu Cartao","Captura da Fatura","Sucesso",$chargeInvoice->invoice_id);
   return array(
     "status" => "success",
     "transid" => $chargeInvoice->invoice_id,
     "rawdata" => $chargeInvoice,
 );
 }
 else {
   logModuleCall("Iugu Cartao","Captura da Fatura","Falhou",$chargeInvoice);
   return array(
    "status" => "declined",
    "rawdata" => $chargeInvoice,
);
 }

} # iugu_cartao_capture


?>
